[macOS] Disable suspension when local notifications are granted

cdumez opened the pull request:
#### 0789169e66412b631fddd23369aa42a65592721a
<pre>
[macOS] Disable suspension when local notifications are granted
<a href="https://bugs.webkit.org/show_bug.cgi?id=256095">https://bugs.webkit.org/show_bug.cgi?id=256095</a>
rdar://107012193

Reviewed by Geoffrey Garen.

On macOS, we now disable process suspension if a tab using this process is likely
to use local notifications. In particular, we disable suspension if either:
- A notification has been shown
- The page called navigator.permissions.query({ name: &quot;notifications&quot; }) and it returned true.
- The page accessed Notification.permission and it returned &quot;granted&quot;.
- The page requested permission via Notification.requestPermission() and it was granted.

This gets reset whenever a new main frame load commits inside the page.

Before this change, our behavior was to restart the 8 minute suspension timer whenever
a local notification was shown.

* Source/WebCore/Modules/permissions/PermissionName.h:
* Source/WebCore/Modules/permissions/PermissionName.idl:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::toPermissionName):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::pageWasNotifiedOfNotificationPermission):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ProcessActivityState::takeLikelyToUseNotificationsActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropLikelyToUseNotificationsActivity):
(WebKit::WebPageProxy::ProcessActivityState::reset):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::queryPermission):
(WebKit::WebPageProxy::pageWillLikelyUseNotifications):
(WebKit::WebPageProxy::showNotification):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::sendMessage):
(WebKit::sendNotificationMessage):
(WebKit::sendNotificationMessageWithAsyncReply):
(WebKit::WebNotificationManager::policyForOrigin const):
(WebKit::WebNotificationManager::sendNotificationMessage): Deleted.
(WebKit::WebNotificationManager::sendNotificationMessageWithAsyncReply): Deleted.
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::checkPermission):

Canonical link: <a href="https://commits.webkit.org/263565@main">https://commits.webkit.org/263565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec391ad292b8a2e25260249d3a2ab9e149eced1b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5081 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5212 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6601 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5168 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5409 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5187 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/5411 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5254 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4556 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6617 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2741 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4554 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9590 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4604 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/4627 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6233 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4141 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4529 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1215 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8608 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4894 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

salinas-miguel commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -485,11 +485,33 @@ void WebPageProxy::ProcessActivityState::takeCapturingActivity()
+void WebPageProxy::ProcessActivityState::takeLikelyToUseNotificationsActivity()
+{
+    if (!m_likelyToUseNotificationsActivity)
+        m_likelyToUseNotificationsActivity = m_page.process().throttler().backgroundActivity("View is likely to use notifications"_s).moveToUniquePtr();
```
If this results in `ProcessThrottler` taking a `WebKit:Background` assertion this may be an issue. `WebKit:Background` is invalidated by RunningBoard after 30 seconds. On macOS we should probably use `WebKit:IndefiniteBackground` which was recently added by RunningBoard

cdumez replied:
Oh, that's good to know. Our background assertion on macOS should really be a IndefiniteBackground it seems.

cdumez replied:
Hopefully addressed via https://github.com/WebKit/WebKit/pull/13278/files#r1180784260.

cdumez commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -321,7 +321,12 @@ - (void)assertion:(RBSAssertion *)assertion didInvalidateWithError:(NSError *)er
     case ProcessAssertionType::Background:
+#if PLATFORM(MAC)
+        // The backgroung assertions timeout after 30 seconds on iOS but not macOS.
+        return @"IndefiniteBackground";
```
@salinas-miguel How about now?

geoffreygaren commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -485,11 +485,33 @@ void WebPageProxy::ProcessActivityState::takeCapturingActivity()
 }
 
+#if PLATFORM(MAC)
+// On macOS, we opt out pages out of process suspension if they used the Notification API to avoid breaking
```
out pages out

cdumez replied:
You're letting me choose which `out` to keep, risky! :D

geoffreygaren reviewed (approved):
r=me

cdumez committed:
[macOS] Disable suspension when local notifications are granted
https://bugs.webkit.org/show_bug.cgi?id=256095
rdar://107012193

Reviewed by Geoffrey Garen.

On macOS, we now disable process suspension if a tab using this process is likely
to use local notifications. In particular, we disable suspension if either:
- A notification has been shown
- The page called navigator.permissions.query({ name: "notifications" }) and it returned true.
- The page accessed Notification.permission and it returned "granted".
- The page requested permission via Notification.requestPermission() and it was granted.

This gets reset whenever a new main frame load commits inside the page.

Before this change, our behavior was to restart the 8 minute suspension timer whenever
a local notification was shown.

* Source/WebCore/Modules/permissions/PermissionName.h:
* Source/WebCore/Modules/permissions/PermissionName.idl:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::toPermissionName):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::pageWasNotifiedOfNotificationPermission):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ProcessActivityState::takeLikelyToUseNotificationsActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropLikelyToUseNotificationsActivity):
(WebKit::WebPageProxy::ProcessActivityState::reset):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::queryPermission):
(WebKit::WebPageProxy::pageWillLikelyUseNotifications):
(WebKit::WebPageProxy::showNotification):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::sendMessage):
(WebKit::sendNotificationMessage):
(WebKit::sendNotificationMessageWithAsyncReply):
(WebKit::WebNotificationManager::policyForOrigin const):
(WebKit::WebNotificationManager::sendNotificationMessage): Deleted.
(WebKit::WebNotificationManager::sendNotificationMessageWithAsyncReply): Deleted.
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp:
(WebKit::WebNotificationClient::checkPermission):

Canonical link: https://commits.webkit.org/263565@main
